src/fs/data_compaction.cc

```diff
@@ -41,7 +41,7 @@
 namespace seastar::fs {
 
 namespace {
-logger mlogger("fs_compaction");
+logger mlogger("fs_data_compaction");
 } // namespace
 
 future<> data_compaction::compact() {
```

Comment:
Consider rewriting this whole function with a seastar thread (`seastar::async`). This process is long and is not hot path, so we can afford allocating a stack for it, and it will be helpful for the following reasons:
 * clearer code - no more tricks like
```cpp
        auto metadata_log_flush = _was_metadata_log_modified ? _metadata_log.flush_log() : now();
        return metadata_log_flush.then([this] {
```
    , no more `do_with()`, `finally()` blocks, `handle_exception()` continuations, etc.
 * easier lifetime management, since everything will be guaranteed to be on our stack
 * better for latencies - for instance, there's a raw `for` loop below in line 144 (`for (auto& file_data_vec : read_data_vecs) {`), and it could be a potential source of stalls if the vector is large. Within `seastar::async` thread, we could just call `thread::maybe_yield()` after each iteration and be done with latency issues

The transformation is really easy, you just:
1. wrap everything in a huge
```cpp
return seastar::async([] {
    (...)
});
```
2. instead of using `then`, just "wait" for futures with `.get()` or `get0()` - seastar thread context will make sure to yield if the future is not ready
3. translate all `finally()` and `handle_exception_type()` to `try {} catch {}`
4. Feel free to translate `read_data_vectors`, `group_data_into_clusters` and other helper functions to regular functions which return void, just add a comment inside them that they are expected to be called only from threaded context. At the beginning of these function, you can assert that by calling `assert(thread::running_in_thread());`

Scylla's compaction process (and many more processes which are long and not on the hot path) are written with seastar light threads and it makes the code way more manageable and error-free.